### DISCUSSION TITLE
How to get pop-up attributes from elements

### DISCUSSION THREAD

**Discussion by immartyn** [pymupdf/PyMuPDF#3189] [Q&A] [answered]
Need help here. For schematic that designed by like Kicard or other CAD tools, the exported PDF has pop-up attributes on the elements like IC, resistors, capacitors, or even wiring lines when you click on them in PDF reader. How can I get these pop-up attributes information of each element(that has pop-up attributes) if processing by PyMuPDF? Thanks!
![kicard_sch](https://github.com/pymupdf/PyMuPDF/assets/28375698/2b5467ca-5de1-4813-8625-d21058a893e4)

**Comment by JorjMcKie** [comment]
Please provide an example page, so we can see what those pop-up are technically.

**Reply by immartyn** [reply]
[audio-adapter-plus-sch.pdf](https://github.com/pymupdf/PyMuPDF/files/14359329/audio-adapter-plus-sch.pdf)
Uploaded a schematic PDF file with more attributes in pop-up window when click on components.
[pic_programmer.pdf](https://github.com/pymupdf/PyMuPDF/files/14360184/pic_programmer.pdf)(another PDF that wires/lines are clickable with pop-up attributes)
<img width="708" alt="image" src="https://github.com/pymupdf/PyMuPDF/assets/28375698/76b4fde9-6bd6-446d-b211-f1c69a67459a">

**Comment by JorjMcKie** [comment]
Thanks for letting us have an example file.
These popups are implemented as links that do not point to another location, but instead they invoke some JavaScript from within some place in the PDF, that does the actual display.

This type of link is not supported by PyMuPDF, neither are there any plans to do so.
But you could achieve something similar with a text annotation. They are show with some icon and display arbitrary text when hovering over them.

**Reply by immartyn** [reply]
Thanks for checking, @JorjMcKie !
Indeed I got nothing from page.get_links() method;
I also didn’t get things from page.annots(). I can only get text annotations the words showed on page with page.get_text(). 
About you said that achieving this similar with a text annotation, can you please elaborate more how to implement? My purpose is to get all the information pop-up window.

**Reply by JorjMcKie** [reply]
Example code:
```python
import fitz

content = """Content can be any UTF-8 string.
Line breaks are respected!
Japanese: コレステロール分子自体は
Hindi: विकिपीडिया सभी"""
doc = fitz.open()
page = doc.new_page()
point = fitz.Point(72, 72)
page.add_text_annot(point, content)
a = page.add_text_annot(point + (0, 30), "The symbol can also be chosen")
a.set_name("Tag")
a.update()

doc.ez_save(__file__.replace(".py", ".pdf"))
```
Looks like this. Text appears upon hovering over the symbol.
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/6fb2dcee-e681-43b0-8489-289b7868b2ce)

**Reply by immartyn** [reply]
This is adding text over the symbol. I mean how to read pop-up text from symbol ?

**Reply by JorjMcKie** [reply]
Sorry - there is currently no way to find out from where the JavaScript is taking its text.

**Reply by immartyn** [reply]
Thanks @JorjMcKie! Do you know any other python modules can do this?

**Reply by JorjMcKie** [reply]
@immartyn there is a misconception here!
We **_can_** find all JavaScript code in the PDF! The question is what next once you have it. For example 
```python
In [1]: import fitz
In [2]: doc=fitz.open("audio-adapter-plus-sch.pdf")
In [3]: print(doc.xref_object(2087))
<<
  /S /JavaScript
  /JS (function ShowCompProps_046929C6977F4B25B9C420B4925DFD9F\(\){var sChoice = app.popUpMenu\(".Creator: DigilentUN",".ReviewerPCB: DigilentUN",".ReviewerSCH: DigilentUN","Alternates: No","Comment: Logo_Chinese ROHS","Component Kind: Standard \(No BOM\)","Description: Chinese ROHS Symbol","DPN: ","Footprint: LOGO_CHINESE-ROHS-2","Ibis Model: ","Library Name: DigilentMaster.IntLib","Library Reference: Logo_Chinese ROHS","LoadStatus: Load","Manufacturer: NA","PCB3D: ","Pin Info: ","Signal Integrity: ","Simulation: ","Value: "\); if \(sChoice\) {var nUrlPos = sChoice.search\(/:\\s\(\(http|https|ftp\):\\/\\/\).+/\); if \(nUrlPos !== -1\) {app.launchURL\(sChoice.substr\(nUrlPos+2\), true\);} else {var nPathPos = sChoice.search\(/:\\s\(\(\\/\)|\(\\\\\\\\\)\).+/\); if \(nPathPos !== -1\) app.openDoc\(sChoice.substr\(nPathPos+2\)\);};};})
>>
```
Would that help you?
The following snippet locates all JavaScript objects and stores the source somewhere as a text file.
```python
In [5]: for xref in range(1,doc.xref_length()):
   ...:     src = doc.xref_get_key(xref, "JS")
   ...:     if src[0] != "string":  # not a JavaScript source object
   ...:         continue
   ...:     out=open(f"{xref}.js", "w")
   ...:     out.write(src[1])
   ...:     out.close()
```
The generated source file "2087.js" then has the following content:
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/1a735daf-565f-4aa4-8ec7-73e464624ba7)

**Comment by JorjMcKie** [comment, marked as answer]
That other file, `pic_programmer.pdf` does not store the JS in separate objects but leaves them inside the abused link objects. So your code must be like this:
```python
for xref in range(1, doc.xref_length()):
    subtype = doc.xref_get_key(xref, "Subtype")
    if subtype[0] != "Link":    # not a Link annotation
        continue
    source = doc.xref_get_key(xref, "A/JS")  # read the JavaScript source inside Link object
    # then do as above
```

**Reply by immartyn** [reply]
This is great, thanks @JorjMcKie! Glad that pymupdf can do this!
